### PR TITLE
Add the calibration tool and document for int8 inference on Paddle

### PR DESCRIPTION
**Pull request by guomingz**
We developed the calibration tool for int8 inference.
Here is the document for this tool.

# Introduction of Accuracy Calibration Tool for 8 Bit Inference
The 8-bits (**INT8**) inference is also known as the Low Precision Inference which could speed up the inference with the lower accuracy loss. It has higher throughput and lower memory requirements compared to FP32. As the PaddlePaddle enables the INT8 inference supporting, we release a accuracy tool(Calibration.py) at the same time. This tool will generate the  quantization parameters and quantized model file finally.

## Usage
1. Build the PaddlePaddle with MKLDNN supporting.
2. `cd /path/to/the/model/fluid/PaddleCV/image_classification`
3. `export FLAGS_use_mkldnn=True`
4. Run the command `python calibration.py --model=MobileNet  --batch_size=50  --class_dim=1000  --image_shape=3,224,224  --with_mem_opt=True  --use_gpu=False  --pretrained_model=weights/mobilenet --out=quantized_out --algo=KL`
    It will generate the specified model (by the parameter --model) to the output directory "quantized_out". Tht tool also provide the parameter "--algo" for KL divergence algorithm which will improve the accuracy.
5. Run the INT8 inference with this command `python eval_int8.py  --iterations=1000  --batch_size=50  --class_dim=1000  --image_shape=3,224,224  --with_mem_opt=True  --use_gpu=False --pretrained_model=quantized_out`. 

## Result
1. Accuracy

| Topology | FP32 Accuracy(Top-1/Top-5) | INT8 Accurary(Top-1/Top-5) |
| --- | :---: | :---: |
ResNet-50 (FB) | 76.63%/93.10%| 76.42%/93.07% 
MobileNet-V1 | 70.78%/89.69% | 70.10%/89.30%

2. Performance

| Topology | FP32 Throughput  | INT8 Throughput | FP32 latency(1x1)  | INT8 latency(1x1) |
| --- | :---: | :---: | :---: | :---: |
ResNet-50 (FB) | 260 | 527 | 62.4 |36.7 
MobileNet-V1 | 1207 |1934 | 9.3 | 8.4

Note: The above performance measured on SKX8180 1S (HT On, Turbo On) 
(Throughput imgs/sec; Latency: ms; 1x1: batch size 1 x thread 1)

### CODE REVIEW

**Comment by hshen14** [comment]
@Superjomn @luotao1 @panyx0718 Please review.

**Comment by luotao1** [comment]
Since the calibration tool for int8 inference is general, is this PR in paddle repo more suitable than in models repo?

**Comment by hshen14** [comment]
@luotao1, we put the calibration tool, following INT8 training-aware solution. As my understanding, it is more suitable to be here, because calibration tool requires related evalution/predict script, which is specific to models, e.g., different eval.py for image classification/object detection/ocr. In the future, as we discussed before, we would like to propose a PASS-based quantization tool, focusing FP32 program to INT8 program transformation; however, how to trigger calibration is still preferrable from Python inferface for more user-friendly data processing.

**Comment by tensor-tang** [comment]
We thought int8 should be a feature of Paddle itself, not relevant with models. And it's better if this Calibration Tool would be placed at https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/scripts and adding with some proper unit tests.

File: fluid/PaddleCV/image_classification/calibration.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    exe = fluid.Executor(place)
+    exe.run(fluid.default_startup_program())
+
+    if pretrained_model:
+        def if_exist(var):
+            return os.path.exists(os.path.join(pretrained_model, var.name))
+
+        fluid.io.load_vars(exe, pretrained_model, predicate=if_exist)
+
+    t = fluid.transpiler.InferenceTranspiler()
```
Comment (panyx0718) [inline fluid/PaddleCV/image_classification/calibration.py:102]: Is this needed?

**Comment by ddokupil** [comment]
I tried running calibration.py and it looks like it changes the image file extension to lowercase. I'm not sure how is that possible but I'm getting following errors:

```
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/paddle/reader/decorator.py", line 281, in handle_worker
    r = mapper(sample)
  File "/data/PaddlePaddle/sources/executor_0/github.com/PaddlePaddle/models/fluid/PaddleCV/image_classification/reader.py", line 102, in process_image
    img = Image.open(img_path)
  File "/usr/lib/python2.7/dist-packages/PIL/Image.py", line 2258, in open
    fp = builtins.open(filename, "rb")
IOError: [Errno 2] No such file or directory: '/data/PaddlePaddle/1G/imagenet/val/ILSVRC2012_val_00000003.jpeg'
```

while my val_list has
```
/data/PaddlePaddle/1G/imagenet/val/ILSVRC2012_val_00000003.JPEG 230
```

When I changed the extension to lowercase for whole dataset I could go a bit further but finally stuck at: 
```
Start calibration for MobileNet...
Traceback (most recent call last):
  File "calibration.py", line 313, in <module>
    main()
  File "calibration.py", line 309, in main
    eval(args)
  File "calibration.py", line 229, in eval
    outputs={"Output": quantize_tmp},
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/framework.py", line 1223, in _insert_op
    op = Operator(block=self, desc=op_desc, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/framework.py", line 578, in __init__
    proto = OpProtoHolder.instance().get_op_proto(type)
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/framework.py", line 482, in get_op_proto
    raise ValueError("Operator \"%s\" has not been registered." % type)
ValueError: Operator "quantize" has not been registered.
```

Also, it'd be convenient if we could point to a custom location of val_list.txt

File: fluid/PaddleCV/image_classification/calibration.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            continue
+        
+        break
+
+    infer_prog = test_program.clone()
+
+    func = ut.get_optimal_scaling_factor if args.algo == 'KL' else np.max
+    for i in conv_input_var_name:
+        ut.update_program_for_saving_var(infer_prog, int8_prog, i+"_scale.input.test",
+         [var_max_range[i]/func(var_max_value_map[i])], np.array(np.max(var_max_value_map[i])).shape, tmp_scale_folder)
```
Comment (panyx0718) [inline fluid/PaddleCV/image_classification/calibration.py:176]: space around `/`

File: fluid/PaddleCV/image_classification/eval_int8.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    program = infer_program.clone()
+    if args.use_transpiler:
+        inference_transpiler_program = program.clone()
+        t = fluid.transpiler.InferenceTranspiler()
+        t.transpile(inference_transpiler_program, place)
+        program = inference_transpiler_program
+
+    fake_data = [(
+        np.random.rand(image_shape[0] * image_shape[1] * image_shape[2]).astype(np.float32),
+        np.random.randint(1, class_dim)) for _ in range(1)]
```
Comment (panyx0718) [inline fluid/PaddleCV/image_classification/eval_int8.py:80]: why loop with `range(1)`?
Should this be put in `if use_fake_data`?

File: fluid/PaddleCV/image_classification/eval_int8.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            data = val_reader().next()
+        image = np.array(map(lambda x: x[0].reshape(image_shape), data)).astype(
+            "float32")
+        label = np.array(map(lambda x: x[1], data)).astype("int64")
+        label = label.reshape([-1, 1])
+
+        t1 = time.time()
+        loss, acc1, acc5 = exe.run(program, feed={feed_dict[0]: image, feed_dict[1]: label}, fetch_list=fetch_targets)
+
+        t2 = time.time()
```
Comment (panyx0718) [inline fluid/PaddleCV/image_classification/eval_int8.py:102]: I'm a bit worried that this line being called everytime will affect performance.
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/executor.cc#L186

File: fluid/PaddleCV/int8/utility.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
+import paddle.fluid.core as core
```
Comment (panyx0718) [inline fluid/PaddleCV/int8/utility.py:1]: core is not part of public API, avoid using it.

File: fluid/PaddleCV/int8/utility.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+    for dot_edge in dot_edges:
+        dot_graph += dot_edge + "\n"
+    for dot_node in dot_nodes:
+        dot_graph += dot_node + "\n"
+    dot_graph += "}"
+
+    with open(output_name, 'w') as f:
+        f.write(dot_graph)
+
+def get_quantize_dequantize_combination(program):
```
Comment (panyx0718) [inline fluid/PaddleCV/int8/utility.py:51]: Please doc the functions, same for others

```
   comments

Args
  ...

Returns
  ...
```

File: fluid/PaddleCV/int8/utility.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+def update_program_attr(program, attr="is_test", value=True):
+    special_op_type_list = [
+                    "pool2d", "sigmoid", "logsigmoid", "softshrink", "exp",
+                    "brelu", "pow", "leaky_relu", "stanh", "relu", "tanh",
+                    "tanh_shrink", "sqrt", "abs", "ceil", "elu", "floor", "cos",
+                    "sin", "round", "reciprocal", "hard_shrink", "hard_sigmoid",
+                    "relu6", "soft_relu", "swish", "thresholded_relu", "log",
+                    "square", "softplus", "softsign"]
+    for i in program.current_block().ops:
+        if i.has_attr(attr) or i.type in special_op_type_list:
```
Comment (panyx0718) [inline fluid/PaddleCV/int8/utility.py:177]: why do you need this special_op_type_list? Are they behaving differently when `is_test=True`?

File: fluid/PaddleCV/int8/utility.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        output_var_name = program.current_block().ops[output_index].output_names[0]
+        output_var = program.current_block().ops[output_index].output(output_var_name)[0]
+        input_var_name = program.current_block().ops[input_index].output_names[0]
+        input_var = program.current_block().ops[input_index].output(input_var_name)[0]
+        quantization_var_name.append(output_var)
+        quantization_var_name.append(input_var)
+    
+    return conv_input_var_name, conv_output_var_name, weights_var_name, quantization_var_name
+
+def check_op_type_by_input(program, var_name, start_index=0):
```
Comment (panyx0718) [inline fluid/PaddleCV/int8/utility.py:327]: why does it want to check? why is `conv2d` `pool2d` special? Maybe the function needs a better name or comment?

**Review by panyx0718 — COMMENTED** [review]
In general, can you add more comments? I will look more into the detail logic.

**Comment by hshen14** [comment]
> We thought int8 should be a feature of Paddle itself, not relevant with models. And it's better if this Calibration Tool would be placed at https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/scripts and adding with some proper unit tests.

Good idea, we will consider.  @tensor-tang @guomingz

**Comment by hshen14** [comment]
> I tried running calibration.py and it looks like it changes the image file extension to lowercase. I'm not sure how is that possible but I'm getting following errors:
> 
> ```
> Exception in thread Thread-4:
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
>     self.run()
>   File "/usr/lib/python2.7/threading.py", line 754, in run
>     self.__target(*self.__args, **self.__kwargs)
>   File "/usr/local/lib/python2.7/dist-packages/paddle/reader/decorator.py", line 281, in handle_worker
>     r = mapper(sample)
>   File "/data/PaddlePaddle/sources/executor_0/github.com/PaddlePaddle/models/fluid/PaddleCV/image_classification/reader.py", line 102, in process_image
>     img = Image.open(img_path)
>   File "/usr/lib/python2.7/dist-packages/PIL/Image.py", line 2258, in open
>     fp = builtins.open(filename, "rb")
> IOError: [Errno 2] No such file or directory: '/data/PaddlePaddle/1G/imagenet/val/ILSVRC2012_val_00000003.jpeg'
> ```
> while my val_list has
> 
> ```
> /data/PaddlePaddle/1G/imagenet/val/ILSVRC2012_val_00000003.JPEG 230
> ```
> When I changed the extension to lowercase for whole dataset I could go a bit further but finally stuck at:
> 
> ```
> Start calibration for MobileNet...
> Traceback (most recent call last):
>   File "calibration.py", line 313, in <module>
>     main()
>   File "calibration.py", line 309, in main
>     eval(args)
>   File "calibration.py", line 229, in eval
>     outputs={"Output": quantize_tmp},
>   File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/framework.py", line 1223, in _insert_op
>     op = Operator(block=self, desc=op_desc, *args, **kwargs)
>   File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/framework.py", line 578, in __init__
>     proto = OpProtoHolder.instance().get_op_proto(type)
>   File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/framework.py", line 482, in get_op_proto
>     raise ValueError("Operator \"%s\" has not been registered." % type)
> ValueError: Operator "quantize" has not been registered.
> ```
> Also, it'd be convenient if we could point to a custom location of val_list.txt

@ddokupil To run INT8, you need below PRs, especially 14464 with INT8 kernel and newly-added ops.

https://github.com/PaddlePaddle/models/pull/1458 
https://github.com/PaddlePaddle/Paddle/pull/14464

BTW, we have the format of validation data for ImageNet:
 val/ILSVRC2012_val_00000001.JPEG 65

File: fluid/PaddleCV/image_classification/calibration.py (hunk 1 above)
Comment (hshen14) [reply fluid/PaddleCV/image_classification/calibration.py:102]: In current implementation, calibration tool reuses the logic from eval/inference script, including data loading, model loading/optimization by transpiler, and inference. So, it is needed here. We have to admit that the code logic will be duplicated somehow, but we want to keep the calibration tool standalone.

File: fluid/PaddleCV/image_classification/calibration.py (hunk 2 above)
Comment (hshen14) [reply fluid/PaddleCV/image_classification/calibration.py:176]: Got it.

File: fluid/PaddleCV/int8/utility.py (hunk 7 above)
Comment (hshen14) [reply fluid/PaddleCV/int8/utility.py:177]: Will remove.

File: fluid/PaddleCV/int8/utility.py (hunk 8 above)
Comment (hshen14) [reply fluid/PaddleCV/int8/utility.py:327]: Right, we need a better name. Basically, conv and pooling are two major primitives running under INT8.

File: fluid/PaddleCV/image_classification/eval_int8.py (hunk 4 above)
Comment (hshen14) [reply fluid/PaddleCV/image_classification/eval_int8.py:102]: We need to invoke the run several times independently:
1. Sampling on FP32: get tensor data, get feeder/fetch vars
2. Saving scale weights for INT8 program

File: fluid/PaddleCV/int8/utility.py (hunk 8 above)
Comment (guomingz) [reply fluid/PaddleCV/int8/utility.py:327]: conv2d & pool2d are the int8-compatible ops at current stage. we need to identify all the ops that input is equal to the specified var name and this op type is int8-compatible or not.

File: fluid/PaddleCV/int8/utility.py (hunk 6 above)
Comment (guomingz) [reply fluid/PaddleCV/int8/utility.py:51]: ok

File: fluid/PaddleCV/image_classification/eval_int8.py (hunk 4 above)
Comment (guomingz) [reply fluid/PaddleCV/image_classification/eval_int8.py:102]: It's a offline(one-shot) tool. The end users won't run it again unless weights was changed.

File: fluid/PaddleCV/image_classification/eval_int8.py (hunk 3 above)
Comment (guomingz) [reply fluid/PaddleCV/image_classification/eval_int8.py:80]: it's just generated the random data for label.

File: fluid/PaddleCV/int8/utility.py (hunk 5 above)
Comment (guomingz) [reply fluid/PaddleCV/int8/utility.py:1]: Do we have an alternative one?

**Comment by luotao1** [comment]
@fc500110 Could you help review this PR?

**Comment by luotao1** [comment]
> Good idea, we will consider

@hshen14 Discussed by @tensor-tang @Superjomn, we think it's better to contribute Calibration Tool to Paddle repo with its unit-test at first. Paddle Repo has CI to ensure the correctness of Calibration Tool, but Models Repo doesn't have it.

**Review by Superjomn — CHANGES_REQUESTED** [review]
Better to contribute to paddle repo directly, the tools in models might be ignored by other paddle developers.

File: fluid/PaddleCV/image_classification/eval_int8.py (hunk 4 above)
Comment (panyx0718) [reply fluid/PaddleCV/image_classification/eval_int8.py:102]: I mean the a problem in executor.cc might affect the time measured here. (Maybe that line is no longer run) Not a problem of this PR
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/executor.cc#L186

Comment (guomingz) [reply fluid/PaddleCV/image_classification/eval_int8.py:102]: got it. Actually, the measured time code is redundant here. We'll remove it in the new PR.

**Comment by ddokupil** [comment]
> @ddokupil To run INT8, you need below PRs, especially 14464 with INT8 kernel and newly-added ops.
> 
> #1458
> [PaddlePaddle/Paddle#14464](https://github.com/PaddlePaddle/Paddle/pull/14464)

@guomingz Are you sure you pasted all PRs needed? I tried with xiaolil1 code and still have issues, only slightly different:

```Start calibration for MobileNet...
Traceback (most recent call last):
  File "calibration.py", line 313, in <module>
    main()
  File "calibration.py", line 309, in main
    eval(args)
  File "calibration.py", line 155, in eval
    max_range = ut.check_op_ancestor(test_program, i.name)
  File "../int8/utility.py", line 128, in check_op_ancestor
    input_var_list.append(op.input(op_input_name)[0])
IndexError: list index out of range
```

**Pull request closed without merging**